### PR TITLE
Try to redo pinnings so that we use run exports

### PR DESCRIPTION
**Pull request by hmaarrfk**
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -55,10 +55,12 @@ outputs:
         - pytorch {{ pytorch }} {{ torch_proc_type }}*
```
Comment (h-vetinari) [inline recipe/meta.yaml:55]: probably want the same in host?
```suggestion
        - pytorch
```

**Review by h-vetinari — COMMENTED** [review]
I like this clean-up, especially as it wasn't very transparent where the jinja-variable `{{ pytorch }}` was coming from. IIUC that's from the migrator (and later, the global pinning), right?

In any case, this is better IMO (barring the adaptation of the host-dep as well).

File: recipe/meta.yaml [hunk 2]
```diff
@@ -55,10 +55,12 @@ outputs:
...
       run:
         - python
-        - pytorch {{ pytorch }} {{ torch_proc_type }}*
+        - pytorch
         - {{ pin_compatible('cudnn') }}          # [cuda_compiler_version != "None"]
         - {{ pin_compatible('numpy') }}
         - pillow >=5.3.0,!=8.3.0,!=8.3.1
+      run_constrained:
+        - pytorch * {{ torch_proc_type }}*
```
Comment (h-vetinari) [inline recipe/meta.yaml:63]: OTOH, should we add a lower bound here based on the version used to build? Not sure if `{{ pin_compatible() }}` would work out of the box, but it should, I guess?

File: recipe/meta.yaml [hunk 3]
```diff
@@ -55,10 +55,12 @@ outputs:
         - pytorch {{ pytorch }} {{ torch_proc_type }}*
       run:
         - python
-        - pytorch {{ pytorch }} {{ torch_proc_type }}*
+        - pytorch
```
Comment (h-vetinari) [inline recipe/meta.yaml:58]: This should do `>= {{ pytorch }}`, AFAIR.
```suggestion
        - {{ pin_compatible("pytorch", max_pin="x") }}
```

File: recipe/meta.yaml (hunk 2 above)
Comment (h-vetinari) [reply recipe/meta.yaml:63]: as a rather irrelevant nit, I dislike significant whitespace. I tend to prefer
```suggestion
        - pytorch =*={{ torch_proc_type }}*
```

File: recipe/meta.yaml (hunk 1 above)
Comment (hmaarrfk) [reply recipe/meta.yaml:55]: Um, i guess we should add the torch_proc_type in the host too. the thing is, we *need* the GPU stuff (as far as I know).

File: recipe/meta.yaml (hunk 3 above)
Comment (hmaarrfk) [reply recipe/meta.yaml:58]: It should be `"x.x"` but we also need to pin the GPU aspect.

Does `pin_compatible` do GPU?

**Comment by hmaarrfk** [comment]
Ok, lets try with the suggestions.

**Comment by hmaarrfk** [comment]
@mikemhenry cc'ing you here to share with you the updates on the pinning.

I guess previously our packages built with 1.10.2 but RAN with 1.10.0 which shouldn't be OK.

File: recipe/meta.yaml (hunk 3 above)
Comment (hmaarrfk) [reply recipe/meta.yaml:59]: We have a run exports in the pytorch package that does this.

File: recipe/meta.yaml (hunk 1 above)
Comment (hmaarrfk) [reply recipe/meta.yaml]: I added a comment in the yaml as to why we can have two constrains in the building sections, but not in the running sections.

File: recipe/meta.yaml (hunk 2 above)
Comment (hmaarrfk) [reply recipe/meta.yaml]: Done.

**Comment by hmaarrfk** [comment]
Thank you for the review @h-vetinari

File: recipe/meta.yaml (hunk 3 above)
Comment (h-vetinari) [reply recipe/meta.yaml:59]: My understanding would be that `pin_compatible` takes care of the lower bound, and the `run_constrained` o the CPU/GPU split

File: recipe/meta.yaml (hunk 1 above)
Comment (h-vetinari) [reply recipe/meta.yaml]: Fine by me. 👍

**Comment by hmaarrfk** [comment]
It is done automatically isn't it?

**Comment by h-vetinari** [comment]
Not that I'm aware. Otherwise, why have all that ceremony (e.g. for numpy)?

**Comment by hmaarrfk** [comment]
When I suggested the solution for numpy, there were other considerations that made it not favorable to implement.

https://github.com/conda-forge/numpy-feedstock/issues/140#issuecomment-469107475

**Comment by hmaarrfk** [comment]
Sigh, ok. the current way doens't do it right:
image

**Comment by hmaarrfk** [comment]
sigh, this is why:
https://github.com/conda-forge/pytorch-cpu-feedstock/pull/89

**Pull request merged**